frontend/model/contracts/mailbox.js

```diff
@@ -1,52 +1,234 @@
 'use strict'
 
 import sbp from '@sbp/sbp'
-import { Vue } from '@common/common.js'
-import { mailType } from './shared/types.js'
-import { objectOf, string, object, optional } from '~/frontend/model/contracts/misc/flowTyper.js'
+import { Vue, L } from '@common/common.js'
+import { merge } from './shared/giLodash.js'
+import { leaveChatRoom } from './shared/functions.js'
+import { object, objectOf, string, optional } from '~/frontend/model/contracts/misc/flowTyper.js'
+import { logExceptNavigationDuplicated } from '~/frontend/views/utils/misc.js'
 
 sbp('chelonia/defineContract', {
   name: 'gi.contracts/mailbox',
   metadata: {
-    // TODO: why is this missing the from username..?
     validate: objectOf({
-      createdDate: string
+      createdDate: string, // action created date
+      username: optional(string), // action creator
+      identityContractID: optional(string) // action creator identityContractID
     }),
     create () {
+      // NOTE: mailbox contract is created before user logs in and also registers either.
+      // so username and identityContractID could be undefined at this time
+      if (!sbp('state/vuex/state').loggedIn) {
+        return { createdDate: new Date().toISOString() }
+      }
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
       return {
-        createdDate: new Date().toISOString()
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
       }
     }
   },
+  getters: {
+    currentMailboxState (state) {
+      return state
+    },
+    ourDirectMessages (state, getters) {
+      return getters.currentMailboxState.dms
+    },
+    ourGroupMessages (state, getters) {
+      return getters.currentMailboxState.gms
+    }
+  },
   actions: {
     'gi.contracts/mailbox': {
-      validate: object, // TODO: define this
+      validate: objectOf({
+        username: string
+      }),
       process ({ data }, { state }) {
+        const initialState = merge({
+          attributes: {
+            creator: data.username,
+            autoJoinAllowance: true // this attribute could be used to block him to be joined direct/group messages automatically by another
+          },
+          dms: {}, // direct message
+          gms: {} // group message
```

Comment:
Shouldn't it be possible to design this so that `dms` and `gms` are the same thing?

Reply:
Sorry, @taoeffect. I don't understand what you mean?
Can you please tell me more in detail? Or taking an example?

Reply:
@Silver-IT let's move this to Slack, I'll reply there